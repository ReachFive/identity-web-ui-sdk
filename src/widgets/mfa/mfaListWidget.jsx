import React from 'react';
import { Info } from '../../components/miscComponent';
import { Card } from '../../components/form/cardComponent';
import { createWidget } from "../../components/widget/widget";
import { withI18n, withTheme } from '../../components/widget/widgetContext'
import styled from "styled-components";

import { ReactComponent as Envelope } from '../../icons/envelope.svg'
import { ReactComponent as CommentAltDots } from '../../icons/comment-alt-dots.svg'

const iconStyle = props => `
    width: ${props.theme.get('fontSize') * 2}px;
    height: ${props.theme.get('fontSize') * 2}px;
    fill: ${props.theme.get('textColor')};
    flex-shrink: 0;
`

const EmailIcon = withTheme(styled(Envelope)`${props => iconStyle(props)}`);
const SmsIcon = withTheme(styled(CommentAltDots)`${props => iconStyle(props)}`);

const credentialIconByType = type => {
    switch (type) {
        case 'email':
            return <EmailIcon />
        case 'sms':
            return <SmsIcon />
    }
}

const Credential = withTheme(styled(Card)`
    display: flex;
    flex-direction: row;
    align-items: center;
    color: ${props => props.theme.get('textColor')};
    line-height: 1.5;
`);

const CardContent = withTheme(styled.div`
    margin-left: ${props => props.theme.get('_blockInnerHeight')}px;
    white-space: initial;
`);

export const MfaList = withI18n((({ credentials, i18n }) => (
    <div>
        {(credentials.length === 0) && (
            <Info>{i18n('mfaList.noCredentials')}</Info>
        )}
        {credentials.map(({type, createdAt, friendlyName, email, phoneNumber}, index) => (
            <Credential key={`credential-${index}`}>
                {credentialIconByType(type)}
                <CardContent>
                    <div style={{fontWeight: 'bold'}}>{friendlyName}</div>
                    <div>{email || phoneNumber}</div>
                    <div>{i18n('mfaList.createdAt')}&nbsp;: <time dateTime={createdAt}>{(new Date(createdAt)).toLocaleDateString()}</time></div>
                </CardContent>
            </Credential>
        ))}
    </div>
)));

export default createWidget({
    component: MfaList,
    prepare: (options, { apiClient }) => {
        return apiClient.listMfaCredentials(options.accessToken).then(credentials => ({
            ...options,
            ...credentials,
        }))
    }
});
