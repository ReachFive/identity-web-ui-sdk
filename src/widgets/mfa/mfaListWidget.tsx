import { MFA } from '@reachfive/identity-core';
import React, { useEffect } from 'react';

import { createWidget } from '../../components/widget/widget';

import { LoaderCircle, Mail, MessageSquareMore, X } from 'lucide-react';
import {
    AlertDialog,
    AlertDialogAction,
    AlertDialogCancel,
    AlertDialogContent,
    AlertDialogDescription,
    AlertDialogFooter,
    AlertDialogHeader,
    AlertDialogTitle,
    AlertDialogTrigger,
} from '../../components/ui/alert-dialog';
import { Button } from '../../components/ui/button';
import { useConfig } from '../../contexts/config';
import { useI18n } from '../../contexts/i18n';
import { useReachfive } from '../../contexts/reachfive.tsx';
import { dateFormat } from '../../helpers/utils.ts';
import { OnError, OnSuccess } from '../../types';

const credentialIconByType = (type: MFA.CredentialsResponse['credentials'][number]['type']) => {
    switch (type) {
        case 'email':
            return <Mail className="bg-background w-icon h-icon stroke-textColor" />;
        case 'sms':
            return <MessageSquareMore className="bg-background w-icon h-icon stroke-textColor" />;
    }
};

export interface MfaListProps {
    /**
     * The authorization credential JSON Web Token (JWT) used to access the ReachFive API, less than five minutes old.
     */
    accessToken: string;
    /**
     * Callback function called when the request has succeeded.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
    /**
     * Indicates whether delete mfa credential button is displayed
     */
    showRemoveMfaCredential?: boolean;
}

interface DeleteButtonProps {
    credential: MFA.Credential;
    isOpen: boolean;
    setIsOpen: React.Dispatch<React.SetStateAction<boolean>>;
    onDeleteCallback: (credential: MFA.Credential) => void;
    deleteConfirmationTitle: string;
    setDeleteConfirmationTitle: React.Dispatch<React.SetStateAction<string>>;
}

const DeleteButton = ({
    credential,
    isOpen,
    setIsOpen,
    onDeleteCallback,
    deleteConfirmationTitle,
    setDeleteConfirmationTitle,
}: DeleteButtonProps) => {
    const i18n = useI18n();

    return (
        <AlertDialog open={isOpen} onOpenChange={setIsOpen}>
            <AlertDialogTrigger asChild>
                <Button
                    variant="destructive"
                    size="icon"
                    className="ml-1"
                    onClick={_ =>
                        setDeleteConfirmationTitle(
                            credential.type === 'email'
                                ? 'mfa.remove.email'
                                : 'mfa.remove.phoneNumber'
                        )
                    }
                >
                    <X />
                </Button>
            </AlertDialogTrigger>
            <AlertDialogContent>
                <AlertDialogHeader>
                    <AlertDialogTitle>{i18n(deleteConfirmationTitle)}</AlertDialogTitle>
                    <AlertDialogDescription />
                </AlertDialogHeader>
                <AlertDialogFooter>
                    <AlertDialogCancel>{i18n('confirmation.cancel')}</AlertDialogCancel>
                    <AlertDialogAction
                        variant="destructive"
                        onClick={_ => onDeleteCallback(credential)}
                    >
                        {i18n('confirmation.yes')}
                    </AlertDialogAction>
                </AlertDialogFooter>
            </AlertDialogContent>
        </AlertDialog>
    );
};

export const MfaList = ({
    accessToken,
    showRemoveMfaCredential,
    onError = (() => {}) as OnError,
    onSuccess = (() => {}) as OnSuccess,
}: MfaListProps) => {
    const [isOpen, setIsOpen] = React.useState(false);
    const [loading, setLoading] = React.useState(true);
    const [deleteConfirmationTitle, setDeleteConfirmationTitle] = React.useState('');
    const [credentials, setCredentials] = React.useState<MFA.Credential[]>([]);
    const i18n = useI18n();
    const config = useConfig();
    const client = useReachfive();

    const fetchMfaCredentials = () => {
        setLoading(true);
        client
            .listMfaCredentials(accessToken)
            .then(mfaCredentialsResponse => {
                setCredentials(mfaCredentialsResponse.credentials);
                onSuccess({
                    name: 'mfa_credentials_listed',
                    credentials: mfaCredentialsResponse.credentials,
                });
            })
            .catch(onError)
            .finally(() => setLoading(false));
    };

    useEffect(() => {
        fetchMfaCredentials();
    }, [accessToken]);

    const onDelete = (credential: MFA.Credential) => {
        switch (credential.type) {
            case 'sms':
                client
                    .removeMfaPhoneNumber({
                        accessToken,
                        phoneNumber: credential.phoneNumber,
                    })
                    .then(() => {
                        fetchMfaCredentials();
                        onSuccess({ name: 'mfa_email_removed' });
                    })
                    .catch(onError);
                break;
            case 'email':
                client
                    .removeMfaEmail({ accessToken })
                    .then(() => {
                        fetchMfaCredentials();
                        onSuccess({ name: 'mfa_email_removed' });
                    })
                    .catch(onError);
                break;
        }
    };

    if (loading) {
        return <LoaderCircle className="animate-spin" />;
    }

    return (
        <div>
            {credentials.length === 0 && (
                <div className="mb-1 text-center text-textColor">
                    {i18n('mfaList.noCredentials')}
                </div>
            )}
            {credentials.map((credential, _) => (
                <div
                    id={`credential-${credential.friendlyName}`}
                    data-testid="credential"
                    key={`credential-${credential.friendlyName}`}
                    className={`flex flex-col ${isOpen ? 'opacity-15' : ''}`}
                >
                    <div className="flex flex-row items-center rounded">
                        <div className="box-border flex flex-row items-center align-middle border border-solid rounded basis-full whitespace-nowrap p-generic">
                            {credentialIconByType(credential.type)}
                            <div className="ml-innerBlock w-max justify-items-stretch text-textColor">
                                <div className="font-bold">{credential.friendlyName}</div>
                                <div>
                                    {MFA.isEmailCredential(credential)
                                        ? credential.email
                                        : MFA.isPhoneCredential(credential)
                                          ? credential.phoneNumber
                                          : 'N/A'}
                                </div>
                                <div>
                                    <span>{i18n('mfaList.createdAt')}&nbsp;: </span>
                                    <time dateTime={credential.createdAt}>
                                        {dateFormat(credential.createdAt, config.language)}
                                    </time>
                                </div>
                            </div>
                        </div>
                        {showRemoveMfaCredential && (
                            <DeleteButton
                                isOpen={isOpen}
                                setIsOpen={setIsOpen}
                                onDeleteCallback={onDelete}
                                deleteConfirmationTitle={deleteConfirmationTitle}
                                setDeleteConfirmationTitle={setDeleteConfirmationTitle}
                                credential={credential}
                            />
                        )}
                    </div>
                </div>
            ))}
        </div>
    );
};

export type MfaListWidgetProps = {
    /**
     * The authorization credential JSON Web Token (JWT) used to access the ReachFive API, less than five minutes old.
     */
    accessToken: string;
    /**
     * Callback function called when the request has succeeded.
     */
    onSuccess?: OnSuccess;
    /**
     * Callback function called when the request has failed.
     */
    onError?: OnError;
    /**
     * Indicates whether delete mfa credential button is displayed
     */
    showRemoveMfaCredential?: boolean;
};

export default createWidget<MfaListWidgetProps, MfaListProps>({
    component: MfaList,
});
