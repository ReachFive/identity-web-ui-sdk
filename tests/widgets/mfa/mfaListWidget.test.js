/**
 * @jest-environment jsdom
 */

import { describe, expect, jest, test } from '@jest/globals';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import mfaListWidget from '../../../src/widgets/mfa/mfaListWidget';

const defaultConfig = { domain: 'local.reach5.net', language: 'fr' };

describe('Snapshot', () => {
    const generateSnapshot = ({ options = {}, config = defaultConfig, credentials }) => () => {
        const apiClient = {
            listMfaCredentials: jest.fn().mockReturnValueOnce(Promise.resolve({ credentials }))
        };

        const tree = mfaListWidget(
            { ...options, accessToken: 'azerty' },
            { config, apiClient }
        ).then(result => renderer.create(result).toJSON());

        expect(tree).resolves.toMatchSnapshot();
    };

    test('empty', generateSnapshot({ credentials: []}))

    test('basic', generateSnapshot({ credentials: [
        { type: 'sms', phoneNumber: '33612345678', friendlyName: 'identifier', createdAt: '2022-09-21' },
        { type: 'email', email: 'root@reach5.co', friendlyName: 'identifier', createdAt: '2022-09-21' }
    ]}))
})
