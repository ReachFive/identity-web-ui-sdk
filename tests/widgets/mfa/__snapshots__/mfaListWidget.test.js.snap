// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot basic 1`] = `
.c1 {
  padding: 9px 38px;
  border: 1px solid #ced4da;
  color: #495057;
  box-sizing: border-box;
  white-space: nowrap;
  vertical-align: middle;
  position: relative;
}

.c1:only-child {
  border-radius: 3px;
}

.c1:not(:only-child):first-child {
  border-radius: 3px 3px 0 0;
}

.c1:not(:only-child):last-child {
  border-radius: 0 0 3px 3px;
}

.c1:not(:only-child):not(:last-child) {
  border-bottom-width: 0;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c5 {
  width: 28px;
  height: 28px;
  fill: #495057;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c3 {
  width: 28px;
  height: 28px;
  fill: #495057;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #495057;
  line-height: 1.5;
}

.c4 {
  margin-left: 38px;
  white-space: initial;
}

<div
  className="c0"
>
  <div>
    <div
      className="c1 c2"
    >
      <svg
        className="c3"
        theme={
          Theme {
            "cache": Object {
              "_absoluteLineHeight": 20,
              "_blockHeight": 40,
              "_blockInnerHeight": 38,
              "animateWidgetEntrance": true,
              "backgroundColor": "#ffffff",
              "borderColor": "#ced4da",
              "borderRadius": 3,
              "borderWidth": 1,
              "fontSize": 14,
              "lineHeight": 1.428571429,
              "maxWidth": 400,
              "paddingY": 9,
              "spacing": 10,
              "textColor": "#495057",
            },
            "defaults": Object {
              "_absoluteLineHeight": [Function],
              "_blockHeight": [Function],
              "_blockInnerHeight": [Function],
              "animateWidgetEntrance": [Function],
              "backgroundColor": [Function],
              "borderColor": [Function],
              "borderRadius": [Function],
              "borderWidth": [Function],
              "button": Object {
                "borderRadius": [Function],
                "borderWidth": [Function],
                "focusBoxShadow": [Function],
                "fontSize": [Function],
                "fontWeight": [Function],
                "height": [Function],
                "lineHeight": [Function],
                "paddingX": [Function],
                "paddingY": [Function],
              },
              "dangerColor": [Function],
              "fontSize": [Function],
              "headingColor": [Function],
              "input": Object {
                "background": [Function],
                "borderColor": [Function],
                "borderRadius": [Function],
                "borderWidth": [Function],
                "boxShadow": [Function],
                "color": [Function],
                "disabledBackground": [Function],
                "focusBorderColor": [Function],
                "focusBoxShadow": [Function],
                "fontSize": [Function],
                "height": [Function],
                "lineHeight": [Function],
                "paddingX": [Function],
                "paddingY": [Function],
                "placeholderColor": [Function],
              },
              "lightBackgroundColor": [Function],
              "lineHeight": [Function],
              "link": Object {
                "color": [Function],
                "decoration": [Function],
                "hoverColor": [Function],
                "hoverDecoration": [Function],
              },
              "maxWidth": [Function],
              "mutedTextColor": [Function],
              "paddingX": [Function],
              "paddingY": [Function],
              "passwordStrengthValidator": Object {
                "color0": [Function],
                "color1": [Function],
                "color2": [Function],
                "color3": [Function],
                "color4": [Function],
              },
              "primaryColor": [Function],
              "smallTextFontSize": [Function],
              "socialButton": Object {
                "borderRadius": [Function],
                "borderWidth": [Function],
                "focusBoxShadow": [Function],
                "fontSize": [Function],
                "fontWeight": [Function],
                "height": [Function],
                "inline": [Function],
                "lineHeight": [Function],
                "paddingX": [Function],
                "paddingY": [Function],
                "textVisible": [Function],
              },
              "spacing": [Function],
              "successColor": [Function],
              "textColor": [Function],
              "warningColor": [Function],
            },
            "props": Object {},
          }
        }
      />
      <div
        className="c4"
      >
        <div
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          identifier
        </div>
        <div>
          33612345678
        </div>
        <div>
          mfaList.createdAt
           : 
          <time
            dateTime="2022-09-21"
          >
            9/20/2022
          </time>
        </div>
      </div>
    </div>
    <div
      className="c1 c2"
    >
      <svg
        className="c5"
        theme={
          Theme {
            "cache": Object {
              "_absoluteLineHeight": 20,
              "_blockHeight": 40,
              "_blockInnerHeight": 38,
              "animateWidgetEntrance": true,
              "backgroundColor": "#ffffff",
              "borderColor": "#ced4da",
              "borderRadius": 3,
              "borderWidth": 1,
              "fontSize": 14,
              "lineHeight": 1.428571429,
              "maxWidth": 400,
              "paddingY": 9,
              "spacing": 10,
              "textColor": "#495057",
            },
            "defaults": Object {
              "_absoluteLineHeight": [Function],
              "_blockHeight": [Function],
              "_blockInnerHeight": [Function],
              "animateWidgetEntrance": [Function],
              "backgroundColor": [Function],
              "borderColor": [Function],
              "borderRadius": [Function],
              "borderWidth": [Function],
              "button": Object {
                "borderRadius": [Function],
                "borderWidth": [Function],
                "focusBoxShadow": [Function],
                "fontSize": [Function],
                "fontWeight": [Function],
                "height": [Function],
                "lineHeight": [Function],
                "paddingX": [Function],
                "paddingY": [Function],
              },
              "dangerColor": [Function],
              "fontSize": [Function],
              "headingColor": [Function],
              "input": Object {
                "background": [Function],
                "borderColor": [Function],
                "borderRadius": [Function],
                "borderWidth": [Function],
                "boxShadow": [Function],
                "color": [Function],
                "disabledBackground": [Function],
                "focusBorderColor": [Function],
                "focusBoxShadow": [Function],
                "fontSize": [Function],
                "height": [Function],
                "lineHeight": [Function],
                "paddingX": [Function],
                "paddingY": [Function],
                "placeholderColor": [Function],
              },
              "lightBackgroundColor": [Function],
              "lineHeight": [Function],
              "link": Object {
                "color": [Function],
                "decoration": [Function],
                "hoverColor": [Function],
                "hoverDecoration": [Function],
              },
              "maxWidth": [Function],
              "mutedTextColor": [Function],
              "paddingX": [Function],
              "paddingY": [Function],
              "passwordStrengthValidator": Object {
                "color0": [Function],
                "color1": [Function],
                "color2": [Function],
                "color3": [Function],
                "color4": [Function],
              },
              "primaryColor": [Function],
              "smallTextFontSize": [Function],
              "socialButton": Object {
                "borderRadius": [Function],
                "borderWidth": [Function],
                "focusBoxShadow": [Function],
                "fontSize": [Function],
                "fontWeight": [Function],
                "height": [Function],
                "inline": [Function],
                "lineHeight": [Function],
                "paddingX": [Function],
                "paddingY": [Function],
                "textVisible": [Function],
              },
              "spacing": [Function],
              "successColor": [Function],
              "textColor": [Function],
              "warningColor": [Function],
            },
            "props": Object {},
          }
        }
      />
      <div
        className="c4"
      >
        <div
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          identifier
        </div>
        <div>
          root@reach5.co
        </div>
        <div>
          mfaList.createdAt
           : 
          <time
            dateTime="2022-09-21"
          >
            9/20/2022
          </time>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot empty 1`] = `
.c1 {
  text-align: center;
  margin-bottom: 10px;
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

<div
  className="c0"
>
  <div>
    <div
      className="c1"
    >
      mfaList.noCredentials
    </div>
  </div>
</div>
`;
